project-koku/koku
# Cost 2312 always log db pid

## Jira Ticket
[COST-2312](https://issues.redhat.com/browse/COST-2312)

## Changes proposed in this PR
* Add DB backend pid logging to log statements and `log_json()` messages
    * This pid will be prepended to the message before logging.
    * The `Logging._log` method is changed to call a new method that will capture the backend pid if available
* Adds new methods `IamTestCase.assertRegexIn` and `IamTestCase.assertRegexNotIn` to help scan iterables such as captured logger output so that checking for messages that may or may not have the backend pid in them.

## Testing Instructions
0. Please for the love of small kittens, run this in `pipenv`
1. Run `python /path/to/manage.py shell`
2. Setup things with these commands
    ```python
    from django.db import connection
    import logging
    import sys

    logging.basicConfig(stream=sys.stdout, force=True, level=logging.INFO)
    LOG = logging.getLogger(__name__)
    ```
3. Execute 
    ```python
    LOG.info("This is my log.")
    ```
    You should see: 
    ```
    INFO:__main__:This is my log
    ````
4. Execute
    ```python
    connection.connect()
    LOG.info("This also is my log")
    ```
    You should see:
    ```
    INFO:__main__:DBPID_<some number> This also is my log
    ```
5. Execute
    ```python
    connection.connection.get_backend_pid()
    ```
    You should see a number matching the `DBPID_` part of the message

---
## Additional Context:

None

## Smoke Tests

```
15:15:00 =========================== short test summary info ============================
15:15:00 FAILED tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_product_search
15:15:00 FAILED tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_projects_compare_data
15:15:00 FAILED tests/rest_api/v1/test_source.py::test_api_gcp_source_raw_calc - Asser...
15:15:00 FAILED tests/rest_api/v1/test_source.py::test_api_ocp_source_crud_upload_service
15:15:00 = 4 failed, 7011 passed, 847 skipped, 1145 deselected, 2 warnings in 3959.29s (1:05:59) =
```

[Console Log](https://ci.ext.devshift.net/job/project-koku-koku-pr-check/3870/console)
